File: packages/go/internal/seguranca/ataques/ataque_cobertura_dependencia_iv.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesCoberturaDependenciaIV configura o AtaqueCoberturaDependenciaIV.
type OpcoesCoberturaDependenciaIV struct {
	TamanhoBloco           int
	PerturbacoesPorPosicao int
}

// AtaqueCoberturaDependenciaIV é o complemento do AtaqueCoberturaDependencia:
// aqui varia o IV, verificando se toda posição do IV influencia toda posição
// de saída.
type AtaqueCoberturaDependenciaIV struct {
	tamanhoBloco           int
	perturbacoesPorPosicao int
}

// NovoAtaqueCoberturaDependenciaIV cria o ataque. Sem argumento, usa
// tamanhoBloco = 32 e perturbacoesPorPosicao = 5.
func NovoAtaqueCoberturaDependenciaIV(opcoes ...OpcoesCoberturaDependenciaIV) *AtaqueCoberturaDependenciaIV {
	o := OpcoesCoberturaDependenciaIV{TamanhoBloco: 32, PerturbacoesPorPosicao: 5}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueCoberturaDependenciaIV{tamanhoBloco: o.TamanhoBloco, perturbacoesPorPosicao: o.PerturbacoesPorPosicao}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueCoberturaDependenciaIV) Nome() string {
	return "Cobertura de dependência do IV (entrada x saída)"
}

// Executar roda o ataque.
func (a *AtaqueCoberturaDependenciaIV) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())
	tamanhoIv := alvo.TamanhoIv()
	ivBase := make([]byte, tamanhoIv)

	ksBase := alvo.GerarKeystreamBruto(chave, ivBase, "enc", a.tamanhoBloco)

	paresIndependentes := []string{}

	for posIv := 0; posIv < tamanhoIv; posIv++ {
		afetou := make([]bool, a.tamanhoBloco)

		for p := 0; p < a.perturbacoesPorPosicao; p++ {
			iv := append([]byte(nil), ivBase...)
			iv[posIv] = byte(seguranca.RandomInt(1, 255))
			ks := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanhoBloco)

			for posSaida := 0; posSaida < a.tamanhoBloco; posSaida++ {
				if ks[posSaida] != ksBase[posSaida] {
					afetou[posSaida] = true
				}
			}
		}

		for posSaida := 0; posSaida < a.tamanhoBloco; posSaida++ {
			if !afetou[posSaida] {
				paresIndependentes = append(paresIndependentes, fmt.Sprintf("iv[%d] -> saida[%d]", posIv, posSaida))
			}
		}
	}

	if len(paresIndependentes) > 0 {
		limite := min(20, len(paresIndependentes))
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevAlta,
			Detalhes: fmt.Sprintf(
				"%d par(es) sem dependência detectável em %d tentativas cada",
				len(paresIndependentes), a.perturbacoesPorPosicao,
			),
			Dados: map[string]any{"pares": paresIndependentes[:limite]},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todas as %d posições do IV influenciam todas as %d posições de saída", tamanhoIv, a.tamanhoBloco),
	}, nil
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueCoberturaDependenciaIV.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Complemento do AtaqueCoberturaDependencia (que varia a CHAVE): aqui varia
 * o IV. Toda posição do IV precisa influenciar toda posição de saída; uma
 * posição "morta" do IV reduziria a entropia efetiva que o IV injeta.
 */
public class AtaqueCoberturaDependenciaIV implements AtaqueInterface {
    private final int tamanhoBloco;
    private final int perturbacoesPorPosicao;

    public AtaqueCoberturaDependenciaIV() {
        this(32, 5);
    }

    public AtaqueCoberturaDependenciaIV(int tamanhoBloco, int perturbacoesPorPosicao) {
        this.tamanhoBloco = tamanhoBloco;
        this.perturbacoesPorPosicao = perturbacoesPorPosicao;
    }

    @Override
    public String nome() {
        return "Cobertura de dependência do IV (entrada x saída)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] chave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8);
        int tamanhoIv = alvo.tamanhoIv();
        byte[] ivBase = new byte[tamanhoIv];

        byte[] ksBase = alvo.gerarKeystreamBruto(chave, ivBase, "enc", this.tamanhoBloco);
        if (ksBase == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        List<String> paresIndependentes = new ArrayList<>();

        for (int posIv = 0; posIv < tamanhoIv; posIv++) {
            boolean[] afetou = new boolean[this.tamanhoBloco];

            for (int p = 0; p < this.perturbacoesPorPosicao; p++) {
                byte[] iv = ivBase.clone();
                iv[posIv] = (byte) Util.randomInt(1, 255);
                byte[] ks = alvo.gerarKeystreamBruto(chave, iv, "enc", this.tamanhoBloco);

                for (int posSaida = 0; posSaida < this.tamanhoBloco; posSaida++) {
                    if (ks[posSaida] != ksBase[posSaida]) {
                        afetou[posSaida] = true;
                    }
                }
            }

            for (int posSaida = 0; posSaida < this.tamanhoBloco; posSaida++) {
                if (!afetou[posSaida]) {
                    paresIndependentes.add("iv[" + posIv + "] -> saida[" + posSaida + "]");
                }
            }
        }

        if (!paresIndependentes.isEmpty()) {
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.ALTA,
                    paresIndependentes.size() + " par(es) sem dependência detectável em "
                            + this.perturbacoesPorPosicao + " tentativas cada",
                    Map.of("pares", paresIndependentes.subList(0, Math.min(20, paresIndependentes.size()))));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "Todas as " + tamanhoIv + " posições do IV influenciam todas as " + this.tamanhoBloco
                        + " posições de saída");
    }
}
